server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -660,6 +721,17 @@ public void moveShards() {
                     if (logger.isTraceEnabled()) {
                         logger.trace("Moved shard [{}] to node [{}]", shardRouting, targetNode.getRoutingNode());
                     }
+                    shardsAllocated++;
+                    // Verifying if this node can be considered ineligible for further iterations
+                    if (targetNode != null) {
+                        Decision nodeLevelAllocationDecision = allocation.deciders().canAllocateAnyShardToNode(
+                            targetNode.getRoutingNode(),
+                            allocation
+                        );
+                        if (nodeLevelAllocationDecision.type() != Decision.Type.YES) {
+                            inEligibleTargetNode.add(targetNode.getRoutingNode());
+                        }
+                    }
```

Comment:
Should we consider abstracting some of the logic here to reduce size.

Reply:
Have abstracted the logic in private method